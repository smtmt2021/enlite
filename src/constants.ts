/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
const isTesting = process.env.NODE_ENV === 'test';
export const TIMEOUT_GET = (isTesting ? 1 : 10) * 1000;
