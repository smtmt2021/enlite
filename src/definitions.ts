/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export const PORT_ENLITE = 3610;
