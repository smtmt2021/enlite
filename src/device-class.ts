/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export enum CLASS {
  NODE_PROFILE = 0x0ef0,
  CONTROLLER = 0x05ff,
  LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER = 0x0288
}

export type ClassType = CLASS.NODE_PROFILE | CLASS.CONTROLLER | CLASS.LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER;
