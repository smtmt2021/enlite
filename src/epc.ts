/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export enum EPC {
   INSTANCE_LIST_NOTIFICATION = 0xd5,

   COEFFICIENT = 0xd3,
   EFFECTIVE_DIGITS = 0xd7,
   UNIT = 0xe1,
   INSTANTANEOUS_ELECTRIC_ENERGY = 0xe7,
   CUMULATIVE_AMOUNTS_OF_ENERGY_NORMAL = 0xea,
   CUMULATIVE_AMOUNTS_OF_ENERGY_REVERSE = 0xeb
 }
