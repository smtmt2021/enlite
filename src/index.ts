/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { Enlite } from './enlite';
import { EnliteImpl } from './enlite-impl';
export { Enlite };
export { SendToRemote as Controller } from './interfaces';
export { Device } from './device';
export { CLASS } from './device-class';
export { SmartElectricEnergyMeter } from './smart-electric-energy-meter';

export function EnliteGetInstance(): Enlite {
  return EnliteImpl.getInstance();
}
