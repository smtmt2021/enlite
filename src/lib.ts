/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { FrameParams, TimeValue } from './interfaces';
import { ESV } from './esv';

/**
 *  Convert number to hex-decimal string.
 *  @param n     to be converted.
 *  @param bytes number of bytes, e.g. 2 bytes means 4 digits.
 */
export function num2hex(n: number, bytes = 2): string {
  if (n > Number.MAX_SAFE_INTEGER) {
    throw new Error(`the specified number is too large, ${n}`);
  }
  if (bytes < 1 || bytes > 8) {
    throw new TypeError(`invalid size of bytes, $${bytes}`);
  }
  return n
    .toString(16)
    .toUpperCase()
    .padStart(bytes * 2, '0');
}

export function dumpFrameParams(params: FrameParams): string {
  let result = '';
  if (params.tid) {
    result += `tid:${num2hex(params.tid, 2)} `;
  }
  if (params.seoj) {
    result += `seoj:${num2hex(params.seoj, 3)} `;
  }
  if (params.deoj) {
    result += `deoj:${num2hex(params.deoj, 3)} `;
  }
  for (const property of params.properties) {
    result += property.toString();
  }
  return result;
}

export function getResponseService(esv: ESV): ESV {
  switch (esv) {
    case ESV.GET:
      return ESV.GET_RES;
    case ESV.INFC:
      return ESV.INFC_RES;
  }
  return esv;
}

export function edt2DateValue(edt?: Buffer): TimeValue | undefined {
  if (!edt || edt.length !== 11) {
    return;
  }
  const year = edt.readUInt16BE(0);
  const month = edt.readUInt8(2) - 1;
  const day = edt.readUInt8(3);
  const hour = edt.readUInt8(4);
  const minuite = edt.readUInt8(5);
  const second = edt.readUInt8(6);
  const value = edt.readUInt32BE(7);
  const time = new Date(year, month, day, hour, minuite, second, 0);
  return { time, value };
}

export function dateValue2Buffer(time: Date, value: number): Buffer {
  const buffer = Buffer.alloc(11);
  buffer.writeUInt16BE(time.getFullYear());
  buffer.writeUInt8(time.getMonth() + 1, 2);
  buffer.writeUInt8(time.getDate(), 3);
  buffer.writeUInt8(time.getHours(), 4);
  buffer.writeUInt8(time.getMinutes(), 5);
  buffer.writeUInt8(time.getSeconds(), 6);
  buffer.writeUInt32BE(value, 7);
  return buffer;
}
