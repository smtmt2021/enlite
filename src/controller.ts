/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { SendToRemote, FrameParams } from './interfaces';
import { Node } from './node';
import { Frame } from './frame';
import { Device } from './device';
import { CLASS } from './device-class';
import DEBUG from 'debug';
const debug = DEBUG('enlite/controller');

export class Controller extends Node implements SendToRemote {
  protected deviceList: Device[] = [];

  constructor(private nodeProfile: SendToRemote) {
    super(CLASS.CONTROLLER);
    debug('controller created');
  }

  async close(): Promise<boolean> {
    debug('controller.close');
    for (const device of this.deviceList) {
      await device.close();
    }
    this.deviceList = [];
    return super.close();
  }

  onData(address: string, frame: Frame): Promise<void> {
    debug('controller.onData', address, frame.seoj.toString(16));
    const device = this.lookup(address, frame.seoj);
    return device ? device.onData(address, frame) : Promise.resolve();
  }

  send(params: FrameParams, address: string): Promise<void> {
    debug('controller.send');
    if (!params.deoj || !params.esv) {
      throw new Error(`Controller.send: params are not enough, ${JSON.stringify(params)}`);
    }
    params.seoj = this.id;
    return this.nodeProfile.send(params, address);
  }

  addDevice(device: Device): Promise<boolean> {
    debug('controller.addDevice');
    if (this.lookup(device.address, device.id)) {
      console.info('enlite: the device is already added', device.id.toString(16), device.address);
      return Promise.resolve(false);
    }
    this.deviceList.push(device);
    return Promise.resolve(true);
  }

  private lookup(address: string, seoj: number): Device | undefined {
    debug('controller.lookup', address, seoj.toString(16));
    return this.deviceList.find(
      device => device.address === address && device.id === seoj
    );
  }
}
