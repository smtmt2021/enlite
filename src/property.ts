/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { EPC } from './epc';
import { num2hex } from './lib';

export class Property {
  static confirmed(property: Property): Property {
    return new Property(property.epc);
  }

  constructor(public epc: EPC, public edt?: Buffer) {}

  get pdc(): number {
    return this.edt ? this.edt.length : 0;
  }

  toString(): string {
    let result = 'epc:' + num2hex(this.epc, 1);
    result += ' pdc:' + num2hex(this.pdc, 1);
    if (this.edt) {
      result += ' edt:';
      for (const edt of this.edt) {
        result += num2hex(edt, 1) + ' ';
      }
    } else {
      result += ' ';
    }
    return result;
  }
}
