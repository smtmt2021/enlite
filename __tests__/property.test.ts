/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { Property } from '../src/property';
import { EPC } from '../src/epc';

test('Property INSTANCE_LIST_NOTIFICATION', () => {
  const property = new Property(EPC.INSTANCE_LIST_NOTIFICATION);
  expect(property.toString()).toMatchSnapshot();
});

test('Property INSTANTANEOUS_ELECTRIC_ENERGY', () => {
  const property = new Property(EPC.INSTANTANEOUS_ELECTRIC_ENERGY);
  expect(property.toString()).toMatchSnapshot();
});

test('Property INSTANCE_LIST_NOTIFICATION with edt', () => {
  const data = Buffer.from('12345678', 'hex');
  const property = new Property(EPC.INSTANCE_LIST_NOTIFICATION, data);
  expect(property.toString()).toMatchSnapshot();
  expect(Property.confirmed(property).toString()).toMatchSnapshot();
});

test('Property INSTANTANEOUS_ELECTRIC_ENERGY with edt', () => {
  const data = Buffer.from('90abcdef12345678', 'hex');
  const property = new Property(EPC.INSTANTANEOUS_ELECTRIC_ENERGY, data);
  expect(property.toString()).toMatchSnapshot();
  expect(Property.confirmed(property).toString()).toMatchSnapshot();
});
