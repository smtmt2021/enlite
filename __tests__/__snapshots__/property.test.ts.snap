// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property INSTANCE_LIST_NOTIFICATION 1`] = `"epc:D5 pdc:00 "`;

exports[`Property INSTANCE_LIST_NOTIFICATION with edt 1`] = `"epc:D5 pdc:04 edt:12 34 56 78 "`;

exports[`Property INSTANCE_LIST_NOTIFICATION with edt 2`] = `"epc:D5 pdc:00 "`;

exports[`Property INSTANTANEOUS_ELECTRIC_ENERGY 1`] = `"epc:E7 pdc:00 "`;

exports[`Property INSTANTANEOUS_ELECTRIC_ENERGY with edt 1`] = `"epc:E7 pdc:08 edt:90 AB CD EF 12 34 56 78 "`;

exports[`Property INSTANTANEOUS_ELECTRIC_ENERGY with edt 2`] = `"epc:E7 pdc:00 "`;
